File: devcontainer/jobportal/commun.py
"""Le peu que les six chapitres partagent."""

from __future__ import annotations

import sys
from pathlib import Path

RACINE = Path(__file__).resolve().parent.parent
sys.path.insert(0, str(RACINE))

AMONT = RACINE / "amont"
FEATURES_AMONT = AMONT / "features"
SCHEMA = AMONT / "devContainer.base.schema.json"
SCHEMA_FEATURE = AMONT / "devContainerFeature.schema.json"
PORTAIL = RACINE / ".devcontainer"
A_CORRIGER = RACINE / "configs" / "a-corriger"


def utf8() -> None:
    """Les chapitres affichent « postCreateCommand » et des guillemets francais.

    Sans ceci, la console Windows (cp1252) leve UnicodeEncodeError au premier
    caractere accentue, et le chapitre meurt avant sa premiere mesure.
    """
    for flux in (sys.stdout, sys.stderr):
        try:
            flux.reconfigure(encoding="utf-8", errors="replace")
        except Exception:      # noqa: BLE001
            pass


def titre(numero: int, texte: str) -> None:
    print(f"\n{numero}. {texte}")
    print("   " + "─" * len(texte))


def ligne(gauche: str, droite: str, largeur: int = 34) -> None:
    print(f"   {gauche:<{largeur}} {droite}")


def tableau(entetes: list[str], lignes: list[list[str]],
            largeurs: list[int]) -> None:
    print("   " + "".join(e.ljust(l) for e, l in zip(entetes, largeurs)))
    print("   " + "".join("─" * (l - 2) + "  " for l in largeurs))
    for rang in lignes:
        print("   " + "".join(str(c).ljust(l) for c, l in zip(rang, largeurs)))


def plier(texte: str, largeur: int = 64) -> list[str]:
    lignes, courante = [], ""
    for mot in texte.split():
        if len(courante) + len(mot) + 1 > largeur:
            lignes.append(courante)
            courante = mot
        else:
            courante = f"{courante} {mot}".strip()
    return lignes + ([courante] if courante else [])

File: n8n/jobportal/commun.js
/** Le peu que les six chapitres partagent. */

'use strict';

function titre(numero, texte) {
  console.log(`\n${numero}. ${texte}`);
  console.log('   ' + '─'.repeat(texte.length));
}

function ligne(gauche, droite, largeur = 34) {
  console.log(`   ${String(gauche).padEnd(largeur)} ${droite}`);
}

function plier(texte, largeur = 64) {
  const lignes = [];
  let courante = '';
  for (const mot of String(texte).split(/\s+/)) {
    if ((`${courante} ${mot}`).trim().length > largeur) {
      lignes.push(courante);
      courante = mot;
    } else courante = (`${courante} ${mot}`).trim();
  }
  if (courante) lignes.push(courante);
  return lignes;
}

/** Un tableau aligné : entêtes puis lignes de cellules. */
function tableau(entetes, lignes, largeurs) {
  const l = largeurs || entetes.map((e) => Math.max(e.length + 2, 12));
  console.log('   ' + entetes.map((e, i) => String(e).padEnd(l[i])).join(''));
  for (const cellules of lignes) {
    console.log('   ' + cellules.map((c, i) => String(c).padEnd(l[i])).join(''));
  }
}

module.exports = { titre, ligne, plier, tableau };
